test/bntest.c

```diff
@@ -1998,29 +2000,37 @@ static int test_rand_range(int n)
         counts[v]++;
     }
 
-    TEST_note("range %u  iterations %u  critical %.4f", range, iterations,
-              critical);
-    if (range < 20) {
-        TEST_note("frequencies (expected %.2f)", expected);
-        for (i = 0; i < range; i++)
-            TEST_note("    %2u  %6zu", i, counts[i]);
-    }
     for (i = 0; i < range; i++) {
         const double delta = counts[i] - expected;
         sum += delta * delta;
     }
     sum /= expected;
-    TEST_note("test statistic %.4f", sum);
 
-    if (TEST_double_lt(sum, critical))
-        res = 1;
+    if (sum > critical) {
+        TEST_info("Chi^2 test negative %.4f > %4.f", sum, critical);
+        TEST_note("test case %zu  range %u  iterations %u", n + 1, range,
+                  iterations);
+        goto err;
+    }
+
+    res = 1;
 err:
     BN_free(rng);
     BN_free(val);
     OPENSSL_free(counts);
     return res;
 }
 
+static int test_rand_range(void)
+{
+    int n_success = 0;
+    size_t i;
+
+    for (i = 0; i < OSSL_NELEM(rand_range_cases); i++)
+        n_success += test_rand_range_single(i);
+    return TEST_int_ge(n_success, binomial_critical);
```

Comment:
Can you also add a higher bound?

Reply:
Out of 50 tests, I had 9 that didn't have any failures ...

Reply:
The upper bound is the end of the distribution.  I.e. currently 37.  The check is for too few successes and only side side is needed for this -- too many successes has no meaning here.

Reply:
No failures on a single run would have a probability of 0.95^37 = 0.14989.  Plugging this into a binomial give a probability of 0.1228.  Low but not unheeded of.

It would be possible to make this two sided test -- I'm not convinced it helps in this case.